File: backend/app/schemas/room.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class CreateRoomRequest(BaseModel):
    """Payload required to create a new coding room."""

    title: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Display name for the room.",
    )
    description: str | None = Field(
        default=None,
        max_length=2000,
        description="Optional longer description of the room's purpose.",
    )
    is_private: bool = Field(
        default=False,
        description="Whether the room requires a password to join.",
    )
    max_participants: int = Field(
        default=10,
        ge=2,
        le=100,
        description="Maximum number of participants allowed (2-100).",
    )

class RoomResponse(BaseModel):
    """Room data returned to the client after creation."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    room_code: str
    title: str
    description: str | None
    host_id: uuid.UUID
    is_private: bool
    max_participants: int
    created_at: datetime


class JoinRoomRequest(BaseModel):
    room_code: str


class ParticipantResponse(BaseModel):
    id: uuid.UUID
    room_id: uuid.UUID
    user_id: uuid.UUID
    is_host: bool

    model_config = ConfigDict(from_attributes=True)


class RoomDetailResponse(BaseModel):
    id: uuid.UUID
    room_code: str
    title: str
    description: str | None
    is_private: bool
    max_participants: int
    participant_count: int
    host_id: uuid.UUID

    model_config = ConfigDict(from_attributes=True)


class LeaveRoomRequest(BaseModel):
    """Request payload for leaving a room."""

    room_code: str = Field(
        ...,
        min_length=8,
        max_length=8,
        pattern=r"^[A-Z0-9]{8}$",
        description="8-character uppercase room code.",
    )

class MessageResponse(BaseModel):
    """Response returned after successfully leaving a room."""

    message: str = Field(
        ...,
        description="Confirmation message.",
    )
    room_code: str = Field(
        ...,
        description="Room that the user left.",
    )
    user_id: uuid.UUID = Field(
        ...,
        description="User who left the room.",
    )

class RoomParticipantResponse(BaseModel):
    """Information about a participant inside a room."""

    user_id: uuid.UUID
    username: str
    is_host: bool
    color: str

    model_config = ConfigDict(from_attributes=True)


class RoomParticipantsResponse(BaseModel):
    """List of all participants in a room."""

    room_code: str
    participant_count: int
    participants: list[RoomParticipantResponse]

    

File: frontend/src/services/room.ts
import api from "./api";

export interface CreateRoomRequest {
  title: string;
  description?: string;
  is_private: boolean;
  max_participants: number;
}

export interface RoomResponse {
  id: string;
  room_code: string;
  title: string;
  description: string | null;
  host_id: string;
  is_private: boolean;
  max_participants: number;
  created_at: string;
}

export async function createRoom(
  data: CreateRoomRequest
): Promise<RoomResponse> {

  const response = await api.post<RoomResponse>(
    "/rooms/create",
    data
  );

  return response.data;
}



export interface RoomDetailResponse {
  id: string;
  room_code: string;
  title: string;
  description: string | null;
  is_private: boolean;
  max_participants: number;
  participant_count: number;
  host_id: string;
}

export async function getRoom(
  roomCode: string
): Promise<RoomDetailResponse> {

  const response = await api.get<RoomDetailResponse>(
    `/rooms/${roomCode}`
  );

  return response.data;
}


export interface JoinRoomRequest {
  room_code: string;
}

export interface ParticipantResponse {
  id: string;
  room_id: string;
  user_id: string;
  is_host: boolean;
}

export async function joinRoom(
  data: JoinRoomRequest
): Promise<ParticipantResponse> {
  const response = await api.post<ParticipantResponse>(
    "/rooms/join",
    data
  );

  return response.data;
}



export interface RoomParticipant {
  user_id: string;
  username: string;
  is_host: boolean;
}

export interface RoomParticipantsResponse {
  room_code: string;
  participant_count: number;
  participants: RoomParticipant[];
}

export async function getParticipants(
  roomCode: string
): Promise<RoomParticipantsResponse> {

  const response = await api.get<RoomParticipantsResponse>(
    `/rooms/${roomCode}/participants`
  );

  return response.data;
}


export interface LeaveRoomRequest {
  room_code: string;
}

export interface LeaveRoomResponse {
  message: string;
  room_code: string;
  user_id: string;
}

export async function leaveRoom(
  roomCode: string
): Promise<LeaveRoomResponse> {

  const response = await api.delete<LeaveRoomResponse>(
    "/rooms/leave",
    {
      data: {
        room_code: roomCode,
      },
    }
  );

  return response.data;
}